DEP: Ensure indexing errors will be raised even on empty results (#15900)
numpy/numpy: doc/release/upcoming_changes/15900.deprecation.rst

Indexing errors will be reported even when selecting no elements
----------------------------------------------------------------
In the future, NumPy will always raise an index error when an
integer array index contains out of bound values.
In complex indexing operations such as::

    arr = np.zeros((5, 5))

>     arr[[], [20]]

Reviewer: Are you sure this example does what you say it does?

This case looks like it should succeed to me, since the semantics are roughly:
```python
index = [], [20]
index = tuple(np.asarray(i, np.intp) for i in index)
index = np.broadcast_arrays(*index)  # 20 is removed here.
arr[index]
```

Author: Yes, which is what we were doing before. The question is whether this is actually nice to ignore the index for checking?

There are two things here pandas had a test for one of them. One is this case, the other is the case below where `arr = np.zeros((5, 5, 0))` but you index only on the first two axis (thus the result also being empty). I could limit on that case I guess...

Maybe this is trickier then I thought.

Author: I guess pandas only ran into the second case, which is easier to reason about... I should probably try to limit this to that one as well...

Reviewer: I think the only way to hit this code path is when the an axis _not_ indexed by advanced indexing is length zero, such as  `np.zeros((3, 0))[[200]]` or `np.zeros((3, 4))[[200], :0]`

Reviewer: Just to be clear - if you run the code sample from this release snippet on this branch, it really does give the warning you claim?

Author: Yeah, its a bit tricky that I check the index before broadcasting.

Indexing errors will be reported even when index result is empty
----------------------------------------------------------------

> In the future, NumPy will raise an index error when an

Reviewer: ```suggestion
In the future, NumPy will raise an IndexError when an
```

> integer array index contains out of bound values but the indexing
> result is empty.

Reviewer: Perhaps
```suggestion
integer array index contains out of bound values even if a non-indexed
dimension is of length 0.
```

> dimension is of length 0.

Reviewer 2: ```suggestion
dimension is of length 0. This will now emit a DeprecationWarning
```

This can happen when the array is previously empty, or an empty

> slice is invloved::

Reviewer: ```suggestion
slice is involved::
```

> slice is involved::

Reviewer: I might be good to give an example of each of these two cases.

The other example could be `arr2 = arr[:, :0]`, `arr2[[20]]`

Author: I applied those doc suggestions, and added an `np.zeros((5, 0))` example

    arr = np.zeros((5, 5))
    arr[[20], :0]

>     # the result: `array([], shape=(1, 0), dtype=float64)` has no elements

Reviewer 2: I think the comment is more confusing than helpful
```suggestion
```

> where no elements are selected. Previously the non-empty index ``[20]``

Reviewer 2: ```suggestion
Previously the non-empty index ``[20]``
```
